components/Heading/Heading.js

```diff
@@ -1,33 +1,84 @@
+import LinkIcon from 'static/images/icons/FontAwesome/link-solid.svg';
 import styles from './Heading.css';
 
-Heading.propTypes = {
-  className: string,
-  id: oneOfType([string, number]), // reference for scroll anchors
-  children: node.isRequired,
-  hasHeadingLines: bool,
-  theme: oneOf(['gray', 'secondary', 'white']),
-};
+class Heading extends Component {
+  static propTypes = {
+    className: string,
+    customAnchorClass: string,
+    text: string.isRequired,
+    hasTitleUnderline: bool,
+    hasHeadingLines: bool,
+    hasHashLink: bool,
+    headingLevel: number,
+    theme: oneOf(['gray', 'secondary', 'white', 'transparentWhite']),
+  };
 
-Heading.defaultProps = {
-  className: undefined,
-  id: '',
-  hasHeadingLines: true,
-  theme: 'gray',
-};
+  static defaultProps = {
+    className: undefined,
+    customAnchorClass: 'anchorMargin',
+    hasTitleUnderline: false,
+    hasHeadingLines: false,
+    hasHashLink: true,
+    headingLevel: 2,
+    theme: 'secondary',
+  };
 
-function Heading({ children, className, hasHeadingLines, id, theme }) {
-  const classes = classNames(className, styles.Heading, styles[theme], {
-    [`${styles.headingLines}`]: hasHeadingLines,
-  });
+  render() {
+    const { props } = this;
+    const anchorId = kebabCase(props.text);
+    const HeadingElement = `h${props.headingLevel}`;
 
-  return (
-    <h2 className={classes} id={id}>
-      {children}
-    </h2>
-  );
+    return (
+      <div className={`${styles.headingContainer}`}>
+        {props.hasHashLink && (
+          <>
+            <span
+              id={anchorId}
+              className={classNames(styles.anchorSpan, styles[props.customAnchorClass])}
+            />
+            <div className={`${styles.hashLinkContainer}`}>
+              <HeadingElement
+                className={classNames(
+                  styles[props.theme],
+                  props.className,
+                  styles.headingTextWithLinkIconOffset,
+                  styles.Heading,
+                )}
+              >
+                <div className={`${styles.hashLink}`}>
+                  <a id={`${anchorId}-link`} href={`#${anchorId}`} data-testid="Hash Link">
+                    <ScreenReaderOnly>Scroll Link for {props.text}</ScreenReaderOnly>
+                    <LinkIcon className={styles.icon} />
+                  </a>
+                  {props.hasTitleUnderline ? (
+                    <span className={`${styles.underline}`}>{props.text}</span>
+                  ) : (
+                    <span>{props.text}</span>
+                  )}
+                </div>
+              </HeadingElement>
+            </div>
+          </>
+        )}
+        {props.hasHeadingLines && (
+          <HeadingElement
+            className={classNames(
+              styles[props.theme],
+              styles[props.className],
+              styles.headingLines,
+            )}
+          >
+            {props.text}
+          </HeadingElement>
+        )}
```

Comment:
I think you can just drop this and put `[props.hasHeadingLines]: styles.headingLines` in the `classNames` function call for the other `<HeadingElement>`

Reply:
The other one is for the HeadingElement that includes the link icon. They are structured differently. The one above it is for the hash link. It needs the span to keep the icon displaying after clicking it. It also requires the link to be wrapped in a div so it can initially be hidden unless hovered over. It also has the link and heading text as children so it won't cause the link icon to appear too far to the left of the heading link when the heading text wraps to another line.

Reply:
If I pass `hasHeadingLines` to a heading component, then it won't have a scroll link.